server/src/main/java/org/eclipse/lsp/cobol/core/engine/CobolLanguageEngine.java

```diff
@@ -125,8 +126,10 @@ public CobolLanguageEngine(
             positionMapping,
             messageService,
             subroutineService);
-    visitor.visit(tree);
-
+    Node rootNode = visitor.visit(tree).get(0);
+    SyntaxTreeEngine syntaxTreeEngine = new SyntaxTreeEngine(rootNode, positionMapping, messageService);
+    syntaxTreeEngine.processTree();
```

Comment:
How is this method heavy (performance point of view)? And runs per opened document or for each update (change) in doc?

Reply:
It's not heavy now. Take a look at times in millis for the biggest source in positive tests:
Preprocessor: 850, parser: 14376, visitor: 4873, SyntaxTreeEngine: 46